File: 1235_maximum_profit_in_job_scheduling/go/main.go
package main

import (
	"fmt"
	"math/rand"
	"sort"
	"time"
)

type Job struct {
	start, end, profit int
}

func maxProfit(start, end, profit []int) int {
	n := len(start)
	jobs := make([]Job, n)
	for i := 0; i < n; i++ {
		jobs[i] = Job{start[i], end[i], profit[i]}
	}

	sort.Slice(jobs, func(i, j int) bool {
		return jobs[i].end < jobs[j].end
	})

	dp := make([]int, n)
	dp[0] = jobs[0].profit

	for i := 1; i < n; i++ {
		left, right := 0, i-1
		include := jobs[i].profit
		for left <= right {
			mid := left + (right-left)>>1
			if jobs[mid].end <= jobs[i].start {
				include = max(include, dp[mid]+jobs[i].profit)
				left = mid + 1
			} else {
			right = mid - 1
			}
		}
		dp[i] = max(include, dp[i-1])
	}
	return dp[n-1]
}

func main() {
	const N = 1_000_000
	start := make([]int, N)
	end := make([]int, N)
	profit := make([]int, N)

	for i := 0; i < N; i++ {
		s := rand.Intn(1_000_000)
		d := rand.Intn(100) + 1
		start[i] = s
		end[i] = s + d
		profit[i] = rand.Intn(1000) + 1
	}

	startTime := time.Now()
	result := maxProfit(start, end, profit)
	elapsed := time.Since(startTime)

	fmt.Printf("Go: %.2f ms → Lucro: %d\n", float64(elapsed.Microseconds())/1000, result)
}

File: 1235_maximum_profit_in_job_scheduling/rust/src/main.rs
use rand::Rng;
use std::time::Instant; // ESSENCIAL!

#[derive(Clone, Copy)]
struct Job {
    start: i32,
    end: i32,
    profit: i32,
}

fn max_profit(start: &[i32], end: &[i32], profit: &[i32]) -> i32 {
    let n = start.len();
    let mut jobs: Vec<Job> = (0..n)
        .map(|i| Job {
            start: start[i],
            end: end[i],
            profit: profit[i],
        })
        .collect();

    // Ordenar por end
    jobs.sort_by_key(|j| j.end);

    let mut dp = vec![0; n];
    dp[0] = jobs[0].profit;

    for i in 1..n {
        let mut low = 0;
        let mut high = i as i32 - 1;
        let mut last = -1;
        let target = jobs[i].start;

        while low <= high {
            let mid = low + (high - low) / 2;
            if jobs[mid as usize].end <= target {
                last = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        let include = jobs[i].profit + if last != -1 { dp[last as usize] } else { 0 };
        dp[i] = include.max(dp[i - 1]);
    }

    dp[n - 1]
}

fn main() {
    const N: usize = 1_000_000;

    let mut start = vec![0; N];
    let mut end = vec![0; N];
    let mut profit = vec![0; N];

    // CORRETO: rng() ao invés de thread_rng()
    let mut rng = rand::thread_rng();

    for i in 0..N {
        let s: i32 = rng.gen_range(0..1_000_000);
        let d: i32 = rng.gen_range(1..101);
        start[i] = s;
        end[i] = s + d;
        profit[i] = rng.gen_range(1..1001);
    }

    let now = Instant::now();
    let result = max_profit(&start, &end, &profit);
    let elapsed = now.elapsed();

    println!(
        "Rust: {:.2} ms → Lucro: {}",
        elapsed.as_secs_f64() * 1000.0,
        result
    );
}
